src/main/java/com/icthh/xm/ms/entity/security/access/DynamicPermissionCheckService.java

```diff
@@ -0,0 +1,154 @@
+        this.permissionCheckService = permissionCheckService;
+    }
+
+    /**
+     * Checks if user has permission with dynamic key feature
+     * if some feature defined by FeatureContext in tenantConfigService enabled TRUE,
+     * then check by @checkContextPermission applied P('XXX'.'YYY') otherwise only basePermission evaluated
+     * @param featureContext FUNCTION|CHANGE_STATE
+     * @param basePermission base permission 'XXX'
+     * @param suffix context permission 'YYY'
+     * @return result from PermissionCheckService.hasPermission
+     */
+    public boolean checkContextPermission(FeatureContext featureContext, String basePermission, String suffix) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(basePermission));
+        Preconditions.checkArgument(StringUtils.isNotEmpty(suffix));
+        if (featureContext.isEnabled(this)) {
+            return checkContextPermission(basePermission, suffix);
+        }
+        return assertPermission(basePermission);
+    }
+
+    /**
+     * Checks if user has permission with dynamic key feature permission = basePermission + "." + suffix
+     * @param basePermission - base permission
+     * @param suffix - suffix
+     * @return result result from PermissionCheckService.hasPermission(permission) from assertPermission
+     */
+    public boolean checkContextPermission(String basePermission, String suffix) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(basePermission));
+        Preconditions.checkArgument(StringUtils.isNotEmpty(suffix));
+        final String permission = basePermission + "." + suffix;
+        return assertPermission(permission);
+    }
+
+    /**
+     * Performs object field adjustment if feature is enabled
+     * @param mapper - adjustment implementation
+     * @param <T> - collection element
+     * @return same collection or adjusted one
+     */
+    public <T> Function<List<T>, List<T>> dynamicFunctionFilter(Function<T, T> mapper) {
+        if (FeatureContext.FUNCTION.isEnabled(this)) {
+            return list -> list.stream().map(mapper).collect(Collectors.toList());
+        }
+        return list -> list;
+    }
+
+    /**
+     * Assert permission via permissionCheckService.hasPermission
+     * @param permission
+     * @return
+     */
+    protected boolean assertPermission(final String permission) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(permission));
+        final boolean permitted = assertPermission.apply(permissionCheckService, permission).booleanValue();
+        if (!permitted) {
+            //TODO place correct error here
+            throw new IllegalStateException("Throw correct exception here");
+        }
+        return Boolean.TRUE.booleanValue();
+    }
+
+    /**
+     * Checks if feature tenant-config -> functions -> dynamic enabled
+     * @return true if feature enabled
+     */
+    private boolean isDynamicFunctionPermissionEnabled() {
```

Comment:
I am against the system transform to a bunch of flags
maybe better to enable permission by default?

Reply:
Enabling something by default is critical security thread.  In live production system you will grant behavior to not privileged users, and this behavior will be available until this grant is not revoked. My strong opinion that enabling functionality by default is good on test, but bad in enterprise environment. @amedvedchuk @ishkurko @vkirichenko @sergeysenja1992 - please discus internally, as standard grant approach.

Reply:
@sergeysenja1992 
Actually as far as i see this flag is needed for backward compatibility. It applies new logic introduced by this service after enabling. In opposite case this logic will be applied by default and some tenants may suffer..